CppImport/src/macro/StaticStuff.cpp

```diff
@@ -0,0 +1,358 @@
+	}
+	else
+		Q_ASSERT(false && "not implemented");
+}
+
+QVector<Model::Node*> StaticStuff::topLevelNodes(QVector<Model::Node*> input)
+{
+	QSet<Model::Node*> topLevel;
+	for (auto node : input) topLevel.insert(node);
+
+	for (auto node : input)
+		for (auto other : input)
+			if (node != other)
+				if (node->isAncestorOf(other))
+					topLevel.remove(other);
+
+	QVector<Model::Node*> result;
+	for (auto node : topLevel) result.append(node);
+	return result;
+}
+
+void StaticStuff::buildMappingInfo(Model::Node* node, QList<Model::Node*>* info)
+{
+	info->push_back(node);
+
+	for (auto child : node->children())
+		buildMappingInfo(child, info);
+}
+
+void StaticStuff::useMappingInfo(Model::Node* node, QList<Model::Node*>* info, NodeMapping* mapping)
+{
+	mapping->add(info->front(), node);
+	info->pop_front();
+
+	for (auto child : node->children())
+		useMappingInfo(child, info, mapping);
+}
+
+OOModel::MetaCallExpression* StaticStuff::containsMetaCall(Model::Node* node)
+{
+	if (auto metaCall = DCast<OOModel::MetaCallExpression>(node))
+		return metaCall;
+
+	for (auto child : node->children())
+		if (auto metaCall = containsMetaCall(child))
+			return metaCall;
+
+	return nullptr;
+}
+
+bool StaticStuff::stringMatches(QString regex, QString value)
+{
+	QRegularExpression regEx(regex);
+	return regEx.match(value).hasMatch();
+}
+
+OOModel::Declaration*StaticStuff::findDeclaration(Model::List* list, QString name)
+{
+	for (auto i = 0; i < list->size(); i++)
+		if (auto decl = DCast<OOModel::Declaration>(list->at(i)))
+			if (decl->name() == name)
+				return decl;
+
+	return nullptr;
+}
+
+OOModel::Expression* StaticStuff::createNameExpressionFromString(QString input)
+{
+	QString baseCase = "((::)?(\\w+(::|\\.|->))*\\w+(\\*|&)?)";
+
+	if (stringMatches("^" + baseCase + "$", input))
+	{
+		QRegularExpression regEx("(::|\\.)");
+		QStringList parts = input.split(regEx);
+
+		QString basePart = parts.last();
+
+		OOModel::Expression* result = nullptr;
+		OOModel::ReferenceExpression* current = nullptr;
+
+		if (basePart.endsWith("&"))
+		{
+			current = new OOModel::ReferenceExpression(basePart.left(basePart.length() - 1));
+			result = new OOModel::ReferenceTypeExpression(current);
+		}
+		else if (basePart.endsWith("*"))
+		{
+			current = new OOModel::ReferenceExpression(basePart.left(basePart.length() - 1));
+			result = new OOModel::PointerTypeExpression(current);
+		}
+		else
+		{
+			current = new OOModel::ReferenceExpression(basePart);
+			result = current;
+		}
+
+		for (auto i = parts.size() - 2; i >= 0; i--)
+		{
+			QString part = parts[i];
+
+			if (part.isEmpty())
+			{
+				Q_ASSERT(i == 0);
+				current->setPrefix(new OOModel::GlobalScopeExpression());
+			}
+			else
+			{
+				auto next = new OOModel::ReferenceExpression(part);
+				current->setPrefix(next);
+				current = next;
+			}
+		}
+
+		return result;
+	}
+	else if (stringMatches("^" + baseCase + "<" + baseCase + ">$", input))
+	{
+		QStringList split = input.split("<");
```

Comment:
Here and in the other places of this method. Instead of manually splitting strings and stuff, which can be very error prone, use the capture groups of the regular expression directly. That way you save work, and your code will look much cleaner.